File: services/mill-metadata-service/src/main/java/io/qpointz/mill/metadata/api/FacetController.java
package io.qpointz.mill.metadata.api;

import io.qpointz.mill.metadata.api.dto.FacetDto;
import io.qpointz.mill.metadata.domain.MetadataEntity;
import io.qpointz.mill.metadata.service.MetadataService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.*;

/**
 * REST controller for scope-aware facet read operations.
 */
@Slf4j
@RestController
@RequestMapping("/api/metadata/v1/facets")
@ConditionalOnBean(MetadataService.class)
@RequiredArgsConstructor
@Tag(name = "Facets", description = "Scope-aware facet management endpoints")
@CrossOrigin(origins = {"http://localhost:5173","http://localhost:8080"})
public class FacetController {
    
    private final MetadataService metadataService;
    
    /**
     * Get merged facet for current user (global + user + team + role).
     */
    @Operation(
        summary = "Get merged facet",
        description = "Retrieves a facet merged from multiple scopes (user > team > role > global) for the specified user context",
        parameters = {
            @Parameter(name = "entityId", description = "Entity ID", required = true, example = "moneta.clients"),
            @Parameter(name = "facetType", description = "Facet type", required = true, example = "descriptive"),
            @Parameter(name = "userId", description = "User ID for user-scoped facets", required = false, example = "alice@company.com"),
            @Parameter(name = "teams", description = "List of team names for team-scoped facets", required = false, example = "engineering"),
            @Parameter(name = "roles", description = "List of role names for role-scoped facets", required = false, example = "admin")
        }
    )
    @ApiResponses({
        @ApiResponse(responseCode = "200", description = "Merged facet found",
            content = @Content(schema = @Schema(implementation = FacetDto.class))),
        @ApiResponse(responseCode = "404", description = "Entity or facet not found")
    })
    @GetMapping("/entities/{entityId}/types/{facetType}")
    public ResponseEntity<FacetDto> getMergedFacet(
        @PathVariable String entityId,
        @PathVariable String facetType,
        @RequestParam(name = "userId", required = false) String userId,
        @RequestParam(name = "teams", required = false) List<String> teams,
        @RequestParam(name = "roles", required = false) List<String> roles
    ) {
        return metadataService.findById(entityId)
            .map(entity -> {
                // Get merged facet
                Optional<Object> mergedData = entity.getMergedFacet(
                    facetType,
                    userId != null ? userId : "anonymous",
                    teams != null ? teams : List.of(),
                    roles != null ? roles : List.of(),
                    Object.class
                );
                
                Set<String> scopes = entity.getFacetScopes(facetType);
                
                return mergedData.map(data -> FacetDto.builder()
                    .facetType(facetType)
                    .data(data)
                    .availableScopes(scopes)
                    .build());
            })
            .filter(Optional::isPresent)
            .map(Optional::get)
            .map(ResponseEntity::ok)
            .orElse(ResponseEntity.notFound().build());
    }
    
    /**
     * Get facet for specific scope.
     */
    @Operation(
        summary = "Get facet by scope",
        description = "Retrieves a facet for a specific scope (e.g., global, user:alice@company.com, team:engineering)",
        parameters = {
            @Parameter(name = "entityId", description = "Entity ID", required = true, example = "moneta.clients"),
            @Parameter(name = "facetType", description = "Facet type", required = true, example = "descriptive"),
            @Parameter(name = "scope", description = "Scope (e.g., global, user:alice@company.com)", required = true, example = "global")
        }
    )
    @ApiResponses({
        @ApiResponse(responseCode = "200", description = "Facet found",
            content = @Content(schema = @Schema(implementation = FacetDto.class))),
        @ApiResponse(responseCode = "404", description = "Entity or facet not found")
    })
    @GetMapping("/entities/{entityId}/types/{facetType}/scopes/{scope}")
    public ResponseEntity<FacetDto> getFacetByScope(
        @PathVariable String entityId,
        @PathVariable String facetType,
        @PathVariable String scope
    ) {
        return metadataService.findById(entityId)
            .map(entity -> {
                Optional<Object> facetData = entity.getFacet(facetType, scope, Object.class);
                Set<String> scopes = entity.getFacetScopes(facetType);
                
                return facetData.map(data -> FacetDto.builder()
                    .facetType(facetType)
                    .data(data)
                    .availableScopes(scopes)
                    .build());
            })
            .filter(Optional::isPresent)
            .map(Optional::get)
            .map(ResponseEntity::ok)
            .orElse(ResponseEntity.notFound().build());
    }
    
}

File: metadata/mill-metadata-service/src/main/kotlin/io/qpointz/mill/metadata/api/FacetController.kt
package io.qpointz.mill.metadata.api

import io.qpointz.mill.metadata.api.dto.FacetDto
import io.qpointz.mill.metadata.service.MetadataService
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.responses.ApiResponse
import io.swagger.v3.oas.annotations.responses.ApiResponses
import io.swagger.v3.oas.annotations.tags.Tag
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.CrossOrigin
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

/** Endpoints for resolved and scope-specific facet retrieval. */
@Tag(name = "Facets", description = "Facet retrieval endpoints")
@RestController
@RequestMapping("/api/metadata/v1/facets")
@CrossOrigin(origins = ["http://localhost:5173", "http://localhost:8080"])
class FacetController(private val metadataService: MetadataService) {

    /** Returns merged facet for requested principal context. */
    @Operation(summary = "Get merged facet for principal context")
    @ApiResponses(
        value = [
            ApiResponse(responseCode = "200", description = "Facet found"),
            ApiResponse(responseCode = "404", description = "Entity or facet not found")
        ]
    )
    @GetMapping("/entities/{entityId}/types/{facetType}")
    fun getMergedFacet(
        @PathVariable entityId: String,
        @PathVariable facetType: String,
        @RequestParam(name = "userId", required = false) userId: String?,
        @RequestParam(name = "teams", required = false) teams: List<String>?,
        @RequestParam(name = "roles", required = false) roles: List<String>?
    ): ResponseEntity<FacetDto> =
        metadataService.findById(entityId)
            .flatMap { entity ->
                val mergedData = entity.getMergedFacet(
                    facetType,
                    userId ?: "anonymous",
                    teams ?: emptyList(),
                    roles ?: emptyList(),
                    Any::class.java
                )
                val scopes = entity.getFacetScopes(facetType)
                mergedData.map { FacetDto(facetType = facetType, data = it, availableScopes = scopes) }
            }
            .map { ResponseEntity.ok(it) }
            .orElse(ResponseEntity.notFound().build())

    /** Returns facet payload for an explicit scope key. */
    @Operation(summary = "Get facet by explicit scope")
    @ApiResponses(
        value = [
            ApiResponse(responseCode = "200", description = "Facet found"),
            ApiResponse(responseCode = "404", description = "Entity or facet not found")
        ]
    )
    @GetMapping("/entities/{entityId}/types/{facetType}/scopes/{scope}")
    fun getFacetByScope(
        @PathVariable entityId: String,
        @PathVariable facetType: String,
        @PathVariable scope: String
    ): ResponseEntity<FacetDto> =
        metadataService.findById(entityId)
            .flatMap { entity ->
                val facetData = entity.getFacet(facetType, scope, Any::class.java)
                val scopes = entity.getFacetScopes(facetType)
                facetData.map { FacetDto(facetType = facetType, data = it, availableScopes = scopes) }
            }
            .map { ResponseEntity.ok(it) }
            .orElse(ResponseEntity.notFound().build())
}
